pkg/resource/role/hooks.go

```diff
@@ -132,3 +134,144 @@ func (rm *resourceManager) detachPolicy(
 	rm.metrics.RecordAPICall("DELETE", "DetachRolePolicy", err)
 	return err
 }
+
+// syncTags examines the Tags in the supplied Role and calls the ListRoleTags,
+// TagRole and UntagRole APIs to ensure that the set of associated Tags  stays
+// in sync with the Role.Spec.Tags
+func (rm *resourceManager) syncTags(
+	ctx context.Context,
+	r *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncTags")
+	defer exit(err)
+	toAdd := []*svcapitypes.Tag{}
+	toDelete := []*svcapitypes.Tag{}
+
+	existingTags, err := rm.getTags(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	for _, t := range r.ko.Spec.Tags {
+		if !inTags(*t.Key, *t.Value, existingTags) {
+			toAdd = append(toAdd, t)
+		}
+	}
+
+	for _, t := range existingTags {
+		if !inTags(*t.Key, *t.Value, r.ko.Spec.Tags) {
+			toDelete = append(toDelete, t)
+		}
+	}
+
+	for _, t := range toAdd {
+		rlog.Debug("adding tag to role", "key", *t.Key, "value", *t.Value)
+	}
+	if err = rm.addTags(ctx, r, toAdd); err != nil {
+		return err
+	}
+	for _, t := range toDelete {
+		rlog.Debug("removing tag from role", "key", *t.Key, "value", *t.Value)
+	}
+	if err = rm.removeTags(ctx, r, toDelete); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// inTags returns true if the supplied key and value can be found in the
+// supplied list of Tag structs.
+//
+// TODO(jaypipes): When we finally standardize Tag handling in ACK, move this
+// to the ACK common runtime/ or pkg/ repos
+func inTags(
+	key string,
+	value string,
+	tags []*svcapitypes.Tag,
+) bool {
+	for _, t := range tags {
+		if *t.Key == key && *t.Value == value {
+			return true
+		}
+	}
+	return false
+}
+
+// getTags returns the list of Policy ARNs currently attached to the Role
+func (rm *resourceManager) getTags(
+	ctx context.Context,
+	r *resource,
+) ([]*svcapitypes.Tag, error) {
+	var err error
+	var resp *svcsdk.ListRoleTagsOutput
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.getTags")
+	defer exit(err)
+
+	input := &svcsdk.ListRoleTagsInput{}
+	input.RoleName = r.ko.Spec.Name
+	res := []*svcapitypes.Tag{}
+
+	for {
+		resp, err = rm.sdkapi.ListRoleTagsWithContext(ctx, input)
+		if err != nil || resp == nil {
+			break
+		}
+		for _, t := range resp.Tags {
+			res = append(res, &svcapitypes.Tag{Key: t.Key, Value: t.Value})
+		}
+		if resp.IsTruncated != nil && !*resp.IsTruncated {
+			break
+		}
```

Comment:
If you are attempting to use pagination for `ListRoleTagsWithContext`, then I think you are missing the `input.Marker` field from this `resp` object.

Reply:
Hmm, yeah. Luckily the limit on tags by default is 50 so in practical matters this won't occur. But, I will fix in a followup PR, thanks for the eagle eyes!